cmd/apiserver/app/server.go

```diff
@@ -0,0 +1,143 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+
+	restfulspec "github.com/emicklei/go-restful-openapi/v2"
+	"github.com/fatih/color"
+	"github.com/go-openapi/spec"
+	"github.com/spf13/cobra"
+
+	"github.com/oam-dev/kubevela/cmd/apiserver/app/options"
+	"github.com/oam-dev/kubevela/pkg/apiserver"
+	"github.com/oam-dev/kubevela/pkg/apiserver/utils/log"
+	"github.com/oam-dev/kubevela/pkg/utils"
+	"github.com/oam-dev/kubevela/version"
+)
+
+// NewAPIServerCommand creates a *cobra.Command object with default parameters
+func NewAPIServerCommand() *cobra.Command {
+	s := options.NewServerRunOptions()
+	cmd := &cobra.Command{
+		Use: "apiserver",
+		Long: `The KubeVela API server validates and configures data for the API objects. 
+The API Server services REST operations and provides the frontend to the
+cluster's shared state through which all other components interact.`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) > 1 && args[0] == "build-swagger" {
+				func() {
+					swagger, err := buildSwagger(s)
+					if err != nil {
+						log.Logger.Fatal(err.Error())
+					}
+					outData, err := json.MarshalIndent(swagger, "", "\t")
+					if err != nil {
+						log.Logger.Fatal(err.Error())
+					}
+					swaggerFile, err := os.OpenFile(args[1], os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0600)
+					if err != nil {
+						log.Logger.Fatal(err.Error())
+					}
+					defer func() {
+						if err := swaggerFile.Close(); err != nil {
+							log.Logger.Errorf("close swagger file failure %s", err.Error())
+						}
+					}()
+					_, err = swaggerFile.Write(outData)
+					if err != nil {
+						log.Logger.Fatal(err.Error())
+					}
+					fmt.Println("build swagger config file success")
+				}()
+				return nil
+			}
+
+			if err := s.Validate(); err != nil {
+				return err
+			}
+
+			return Run(signals.SetupSignalHandler(), s)
+		},
+		SilenceUsage: true,
+	}
+
+	fs := cmd.Flags()
+	namedFlagSets := s.Flags()
+	for _, set := range namedFlagSets.FlagSets {
+		fs.AddFlagSet(set)
+	}
+
+	return cmd
+}
+
+// Run runs the specified APIServer. This should never exit.
+func Run(ctx context.Context, s *options.ServerRunOptions) error {
+	// The server is not terminal, there is no color default.
+	// Force set to false, this is useful for the dry-run API.
+	color.NoColor = false
+
+	errChan := make(chan error)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	if s.GenericServerRunOptions.PprofAddr != "" {
+		go utils.EnablePprof(s.GenericServerRunOptions.PprofAddr, errChan)
+	}
+
+	go func() {
+		if err := run(ctx, s, errChan); err != nil {
+			errChan <- fmt.Errorf("failed to run apiserver: %w", err)
+		}
+	}()
+	var term = make(chan os.Signal, 1)
+	signal.Notify(term, os.Interrupt, syscall.SIGTERM)
```

Comment:
Is it repeated with the `signals.SetupSignalHandler()`?

Reply:
yes, it has been done